pallets/bridge/src/lib.rs

```diff
@@ -0,0 +1,621 @@
+// Copyright 2020-2021 Litentry Technologies GmbH.
+// This file is part of Litentry.
+//
+// Litentry is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod hashing;
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use codec::{Decode, Encode, EncodeLike};
+	pub use frame_support::{
+		pallet_prelude::*, traits::StorageVersion, weights::GetDispatchInfo, PalletId, Parameter,
+	};
+	use frame_system::{
+		pallet_prelude::*,
+		{self as system},
+	};
+	use scale_info::TypeInfo;
+	pub use sp_core::U256;
+	use sp_runtime::{
+		traits::{AccountIdConversion, Dispatchable},
+		RuntimeDebug,
+	};
+	use sp_std::prelude::*;
+
+	const DEFAULT_RELAYER_THRESHOLD: u32 = 1;
+	const MODULE_ID: PalletId = PalletId(*b"litry/bg");
+
+	pub type BridgeChainId = u8;
+	pub type DepositNonce = u64;
+	pub type ResourceId = [u8; 32];
+
+	/// Helper function to concatenate a chain ID and some bytes to produce a resource ID.
+	/// The common format is (31 bytes unique ID + 1 byte chain ID).
+	pub fn derive_resource_id(chain: u8, id: &[u8]) -> ResourceId {
+		let mut r_id: ResourceId = [0; 32];
+		r_id[31] = chain; // last byte is chain id
+		let range = if id.len() > 31 { 31 } else { id.len() }; // Use at most 31 bytes
```

Comment:
I think here a boundary check is missing: `range - 1 - i` can go negative

Reply:
the minimum value of "range - 1 - i" is zero.

Reply:
> the minimum value of "range - 1 - i" is zero.

-1 I think? range can be 0, i can be 0